### PR TITLE
Follow-up to #1276: Remove multiple versions of Slice/Container getCRAIEntries()

### PR DESCRIPTION
**Pull request by jmthibault79**
- Encapsulate Container.byteOffset and distribute to Slices on set
- enforce prerequisites to index entry creation

### Description

Please explain the changes you made here.
Explain the **motivation** for making this change. What existing problem does the pull request solve?

### Checklist

- [x] Code compiles correctly
- [x] New tests covering changes and new functionality
- [x] All tests passing
- [x] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

### CODE REVIEW

File: src/main/java/htsjdk/samtools/BAMIndexMetaData.java [hunk 1]
```diff
@@ -159,7 +159,7 @@ void recordMetaData(final Slice slice) {
             unAlignedRecords += slice.unmappedReadsCount;
         }
 
-        final long start = slice.offset;
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/BAMIndexMetaData.java]: renamed the Slice indexing parameters to make their function clearer.

File: src/main/java/htsjdk/samtools/CRAMBAIIndexer.java [hunk 2]
```diff
@@ -125,14 +124,12 @@ public void processContainer(final Container container, final ValidationStringen
 
         int sliceIndex = 0;
         for (final Slice slice : container.slices) {
-            slice.containerOffset = container.offset;
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/CRAMBAIIndexer.java]: this is now set elsewhere

File: src/main/java/htsjdk/samtools/cram/CRAIIndex.java [hunk 3]
```diff
@@ -59,7 +59,7 @@ public void writeIndex(final OutputStream os) {
      * @param container the container to index
      */
     public void processContainer(final Container container) {
-        addEntries(container.getCRAIEntriesSplittingMultiRef());
+        addEntries(container.getCRAIEntries());
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/CRAIIndex.java]: we always split multiref now

File: src/main/java/htsjdk/samtools/cram/build/ContainerFactory.java [hunk 4]
```diff
@@ -58,8 +58,7 @@ public Container buildContainer(final List<CramCompressionRecord> records) {
             slices.add(slice);
         }
 
-        final Container container = Container.initializeFromSlices(slices);
-        container.compressionHeader = compressionHeader;
+        final Container container = Container.initializeFromSlices(slices, compressionHeader);
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/build/ContainerFactory.java]: We need the compressionHeader when we call `getCraiEntries()` so requiring it's here helps ensure it's valid

File: src/main/java/htsjdk/samtools/cram/structure/Container.java [hunk 5]
```diff
@@ -137,34 +156,16 @@ else if (sliceRefContexts.size() > 1) {
...
     public List<CRAIEntry> getCRAIEntries() {
-        return Arrays.stream(slices)
-                .map(Slice::getCRAIEntry)
-                .collect(Collectors.toList());
-    }
-
-    /**
-     * Retrieve the list of CRAI Index entries corresponding to this Container.
-     *
-     * TODO: investigate why we sometimes split multi-ref Slices
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/structure/Container.java]: now we always split

File: src/main/java/htsjdk/samtools/cram/structure/Container.java [hunk 6]
```diff
@@ -137,34 +156,16 @@ else if (sliceRefContexts.size() > 1) {
...
-     * into different entries and sometimes do not
-     *
-     * TODO: clearly identify and enforce preconditions, e.g.
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/structure/Container.java]: precondition enforcement is in `craiIndexInitializationCheck()`

File: src/main/java/htsjdk/samtools/cram/structure/ContainerIO.java [hunk 7]
```diff
@@ -95,18 +95,20 @@ private static Container readContainer(final int major, final InputStream inputS
     }
 
     private static void calculateSliceOffsetsAndSizes(final Container container) {
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/structure/ContainerIO.java]: Make sure to carefully coordinate this change with #1326

File: src/main/java/htsjdk/samtools/cram/structure/Slice.java [hunk 8]
```diff
@@ -337,51 +372,42 @@ public CramRecordReader createCramRecordReader(final CompressionHeader header,
...
+        craiIndexInitializationCheck();
+
         if (referenceContext.isMultiRef()) {
             final Map<ReferenceContext, AlignmentSpan> spans = getMultiRefAlignmentSpans(compressionHeader, ValidationStringency.DEFAULT_STRINGENCY);
 
             return spans.entrySet().stream()
                     .map(e -> new CRAIEntry(e.getKey().getSerializableId(),
                             e.getValue().getStart(),
                             e.getValue().getSpan(),
-                            containerStartByteOffset,
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/structure/Slice.java]: can trust the Slice now instead of relying on the Container, for these 2 values

File: src/main/java/htsjdk/samtools/cram/structure/Slice.java [hunk 9]
```diff
@@ -337,51 +372,42 @@ public CramRecordReader createCramRecordReader(final CompressionHeader header,
...
-                            landmarks[index],
-                            size))
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/structure/Slice.java]: same member variable.  just renamed.

File: src/main/java/htsjdk/samtools/cram/structure/Slice.java [hunk 10]
```diff
@@ -337,51 +372,42 @@ public CramRecordReader createCramRecordReader(final CompressionHeader header,
...
                     .collect(Collectors.toList());
         } else {
             // single ref or unmapped
             final int sequenceId = referenceContext.getSerializableId();
-            return Collections.singletonList(new CRAIEntry(sequenceId, alignmentStart, alignmentSpan, containerStartByteOffset, offset, size));
+            return Collections.singletonList(new CRAIEntry(
+                    sequenceId,
+                    alignmentStart,
+                    alignmentSpan,
+                    containerByteOffset,
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/structure/Slice.java]: `containerByteOffset` newly uses the Slice member variable.
`byteOffsetFromContainer` and `byteSize` are continuing uses of the Slice member variables, renamed.

**Comment by jmthibault79** [comment]
This is going to conflict with #1326 so let's hold off on review for now, and wait for a rebase.

File: src/main/java/htsjdk/samtools/cram/structure/Container.java [hunk 11]
```diff
@@ -67,11 +67,28 @@
...
+     * @param byteOffset the byte location in the stream where this Container begins
      */
-    public long offset;
+    public void setByteOffset(final long byteOffset) {
+        this.byteOffset = byteOffset;
+        if (slices != null) {
+            for (final Slice slice : slices) {
+                slice.containerByteOffset = byteOffset;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/structure/Container.java]: @jmthibault79 It seems like it should be possible to reduce the number of call sites that need to update these offsets. Every call to this method (except for the one in ContainerStreamWriter, when a new container is being created from records) is immediately preceded by a call (either directly or indirectly), to ContainerIO.readContainer. It would be much  simpler to make ContainerIO responsible for all of those cases, even if it means passing the offset in (although I think it could often derive the offset from the stream position). Would that work ? I think it would reduce the complexity around understanding who/when the slice offsets get propagated, etc.

[the remark this answers is not in the document]
Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/cram/structure/Container.java]: I do think this is possible, and I have started an attempt at this.  I'll work on that some more.

Good point about propagation.  I will have to think more about that.

[the remark this answers is not in the document]
Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/cram/structure/Container.java]: Addressed both of these in a new commit.

File: src/main/java/htsjdk/samtools/CRAMBAIIndexer.java (hunk 2 above)
Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/CRAMBAIIndexer.java]: `Container.setSlices()` sets this value.  If the slices exist prior to the setting of this value, then `Container.setByteOffset()` sets the slices' values.

File: src/main/java/htsjdk/samtools/CRAMCRAIIndexer.java [hunk 12]
```diff
@@ -94,27 +93,19 @@ public void finish() {
...
+        final CramHeader cramHeader = CramIO.readCramHeader(cramStream);
+        final CRAMCRAIIndexer indexer = new CRAMCRAIIndexer(craiStream, cramHeader.getSamFileHeader());
+        final Version cramVersion = cramHeader.getVersion();
 
-            indexer.finish();
-        }
-        catch (IOException e) {
-            throw new RuntimeIOException("Error writing CRAI index to output stream");
+        // get the first container
+        Container container = ContainerIO.readContainer(cramVersion, cramStream);
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/CRAMCRAIIndexer.java]: offset is set inside `readContainer()` by querying the stream

File: src/main/java/htsjdk/samtools/CRAMFileReader.java [hunk 13]
```diff
@@ -400,7 +400,7 @@ public SAMRecordIterator assertSorted(final SortOrder sortOrder) {
             seekableStream.seek(0);
             newIterator = new CRAMIterator(seekableStream, referenceSource, validationStringency);
             seekableStream.seek(startOfLastLinearBin >>> 16);
-            final Container container = ContainerIO.readContainerHeader(newIterator.getCramHeader().getVersion().major, seekableStream);
+            final Container container = ContainerHeaderIO.readContainerHeader(newIterator.getCramHeader().getVersion().major, seekableStream);
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/CRAMFileReader.java]: this was driving me nuts

Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/CRAMFileReader.java]: TODO obviously: move this to a new `ContainerHeader` class.  Just not today.

File: src/main/java/htsjdk/samtools/cram/build/ContainerFactory.java [hunk 14]
```diff
@@ -58,8 +64,7 @@ public Container buildContainer(final List<CramCompressionRecord> records) {
             slices.add(slice);
         }
 
-        final Container container = Container.initializeFromSlices(slices);
-        container.compressionHeader = compressionHeader;
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/build/ContainerFactory.java]: I don't recall why I moved this inside the call to `initializeFromSlices()`.  I can take it back out - they seem pretty equal to me.

File: src/main/java/htsjdk/samtools/cram/build/CramContainerHeaderIterator.java [hunk 15]
```diff
@@ -22,8 +22,19 @@ public CramContainerHeaderIterator(final InputStream inputStream) {
...
+     * Consume the entirety of the next container from the stream,
+     * but retain only the header.
+     *
+     * @see CramContainerIterator#containerFromStream(Version, CountingInputStream)
+     *
+     * @param cramVersion the expected CRAM version of the stream
+     * @param countingStream the {@link CountingInputStream} to read from
+     * @return The next Container's header from the stream, returned as a Container.
+     */
+    @Override
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/build/CramContainerHeaderIterator.java]: This tripped me up!

Hopefully this explicit Override and Javadoc will save some future grief.

File: src/main/java/htsjdk/samtools/cram/build/CramContainerIterator.java [hunk 16]
```diff
@@ -17,20 +17,14 @@
     private CountingInputStream countingInputStream;
     private Container nextContainer;
     private boolean eof = false;
-    private long offset = 0;
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/build/CramContainerIterator.java]: containers can now take care of this.  don't need it here.

File: src/main/java/htsjdk/samtools/cram/build/CramIO.java [hunk 17]
```diff
@@ -293,36 +297,39 @@ private static SAMFileHeader readSAMFileHeader(final Version version, InputStrea
             } else {
                 /*
                  * pending issue: container.containerByteSize inputStream 2 bytes shorter
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/build/CramIO.java]: I have not investigated this

File: src/main/java/htsjdk/samtools/cram/build/CramIO.java [hunk 18]
```diff
@@ -293,36 +297,39 @@ private static SAMFileHeader readSAMFileHeader(final Version version, InputStrea
...
-				 * then needed in the v21 test cram files.
-				 */
+                 * than needed in the v21 test cram files.
+                 */
                 block = Block.read(version.major, inputStream);
             }
         }
 
-        inputStream = new ByteArrayInputStream(block.getUncompressedContent());
+        byte[] bytes;
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/build/CramIO.java]: not really related to this PR.  Just a little cleanup while I'm here.

File: src/main/java/htsjdk/samtools/cram/build/CramSpanContainerIterator.java [hunk 19]
```diff
@@ -99,10 +99,7 @@ public Container next() {
                     throw new RuntimeException("No more containers in this boundary.");
                 }
                 
-                final long offset = seekableStream.position();
-                final Container c = ContainerIO.readContainer(cramHeader.getVersion(), seekableStream);
-                c.offset = offset;
-                return c;
+                return ContainerIO.readContainer(cramHeader.getVersion(), seekableStream);
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/build/CramSpanContainerIterator.java]: `readContainer()` takes care of setting the offset by querying the `SeekableStream`

File: src/main/java/htsjdk/samtools/cram/structure/Container.java [hunk 20]
```diff
@@ -65,13 +65,55 @@
     public CompressionHeader compressionHeader;
 
     // slices found in the container:
-    public Slice[] slices;
+    private Slice[] slices;
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/structure/Container.java]: Can't set the slices without also setting their byteOffsets now

File: src/main/java/htsjdk/samtools/cram/structure/ContainerHeaderIO.java [hunk 21]
```diff
@@ -70,9 +77,42 @@ public static Container readContainerHeader(final int major, final InputStream i
         if (major >= 3)
             container.checksum = CramInt.readInt32(inputStream);
 
+        container.setByteOffset(containerByteOffset);
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/structure/ContainerHeaderIO.java]: this is now the only remaining external call to `setByteOffset()`

File: src/main/java/htsjdk/samtools/cram/structure/ContainerHeaderIO.java [hunk 22]
```diff
@@ -70,9 +77,42 @@ public static Container readContainerHeader(final int major, final InputStream i
...
         return container;
     }
 
+    // convenience methods for SeekableStream and CountingInputStream
+    // TODO: merge these two classes?
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/structure/ContainerHeaderIO.java]: these 2 seem VERY similar.  It's annoying to have to account for both.

Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/cram/structure/ContainerHeaderIO.java]: Maybe Seekable can inherit from Counting?

File: src/main/java/htsjdk/samtools/cram/structure/ContainerIO.java [hunk 23]
```diff
@@ -19,89 +21,99 @@
...
+     * @param inputStream the {@link InputStream} stream to read from
+     * @param containerByteOffset the byte offset from the start of the stream
      * @return a new container object read from the stream
      */
-    public static Container readContainer(final Version version, final InputStream inputStream) {
-        final Container container = readContainer(version.major, inputStream);
+    public static Container readContainer(final Version version,
+                                          final InputStream inputStream,
+                                          final long containerByteOffset) {
+        Container container = readContainerInternal(version.major, inputStream, containerByteOffset);
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/structure/ContainerIO.java]: Another version of this `readContainer()` consolidation is in #1306

File: src/main/java/htsjdk/samtools/cram/structure/ContainerIO.java [hunk 24]
```diff
@@ -19,89 +21,99 @@
...
 
-        final Container container = readContainerHeader(major, inputStream);
+        final Container container = ContainerHeaderIO.readContainerHeader(major, inputStream, containerByteOffset);
         if (container.isEOF()) {
             return container;
         }
 
         container.compressionHeader = CompressionHeader.read(major, inputStream);
 
-        howManySlices = Math.min(container.landmarks.length, howManySlices);
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/structure/ContainerIO.java]: `howManySlices` and `fromSlice` were never used.  See #1306 for more details.

File: src/main/java/htsjdk/samtools/cram/structure/ContainerIO.java [hunk 25]
```diff
@@ -19,89 +21,99 @@
...
 
-    /**
-     * Writes a {@link Container} header information to a {@link OutputStream}.
-     *
-     * @param major     the CRAM version to assume
-     * @param container the container holding the header to write
-     * @param outputStream        the stream to write to
-     * @return the number of bytes written
-     */
-    public static int writeContainerHeader(final int major, final Container container, final OutputStream outputStream) {
```
Comment (jmthibault79) [inline src/main/java/htsjdk/samtools/cram/structure/ContainerIO.java]: why did this exist

File: src/test/java/htsjdk/samtools/CRAMSliceMD5Test.java [hunk 26]
```diff
@@ -30,10 +31,14 @@ public void testSliceMD5() throws IOException {
         final CramTestCase test = new CramTestCase();
 
         // read the CRAM:
-        final ByteArrayInputStream bais = new ByteArrayInputStream(test.cramData);
-        final CramHeader cramHeader = CramIO.readCramHeader(bais);
-        final Container container = ContainerIO.readContainer(cramHeader.getVersion(), bais);
-        final Slice slice = container.slices[0];
+        Container container;
```
Comment (jmthibault79) [inline src/test/java/htsjdk/samtools/CRAMSliceMD5Test.java]: more unrelated misc cleanup

File: src/main/java/htsjdk/samtools/cram/structure/Slice.java [hunk 27]
```diff
@@ -75,16 +75,51 @@
...
-    public int size = -1;
-    // this Slice's index within its Container
-    public int index = -1;
+    // CRAI only
+    public int byteSize = UNINITIALIZED_INDEXING_PARAMETER;
+    // this Slice's index number within its Container
+    // BAI only
+    public int index = UNINITIALIZED_INDEXING_PARAMETER;
+
+    void baiIndexInitializationCheck() {
```
Comment (cmnbroad) [inline src/main/java/htsjdk/samtools/cram/structure/Slice.java]: This looks like a good check, but it doesn't appear to be called from anywhere except within one test (`craiIndexInitializationCheck` below *is* called before indexing).

Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/structure/Slice.java]: I started looking at this a bit more. It looks like `populateSlicesAndIndexingParameters` is only called when we're creating an index for a CRAM file that we've read in from disk, but I'm unclear on how the slice index info gets populated in the case where we're creating the index on-the-fly, when writing a new CRAM file. Since the tests were passing, I assumed it was happening somehow. So, I naively tried adding a call to this method from within CRAMBAIIndexer.processContainer, and then ran as many of the indexing tests as I easily could. 6 of them failed. So this led to a few questions:
- Where should the call to `baiIndexInitializationCheck` method go ?
- Does the equivalent of `populateSlicesAndIndexingParameters` happen in the case where we're writing a cram out from SAMRecords ?
- If we aren't populating the slice index info when writing a new CRAM file on this branch, how/why is it that all of the index tests are still passing ?
- Is there a similar issue already on master that isn't being caught by tests ?

File: src/main/java/htsjdk/samtools/cram/build/ContainerFactory.java [hunk 28]
```diff
@@ -37,7 +37,13 @@ public ContainerFactory(final SAMFileHeader samFileHeader, final int recordsPerS
         this.recordsPerSlice = recordsPerSlice;
     }
 
-    public Container buildContainer(final List<CramCompressionRecord> records) {
+    /**
+     *
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/cram/build/ContainerFactory.java]: Oops, yeah

File: src/main/java/htsjdk/samtools/cram/structure/ContainerIO.java [hunk 29]
```diff
@@ -19,89 +21,97 @@
...
-        try {
-            if (fromSlice > 0) //noinspection ResultOfMethodCallIgnored
-                inputStream.skip(container.landmarks[fromSlice]);
-        } catch (final IOException e) {
-            throw new RuntimeIOException(e);
-        }
-
         final ArrayList<Slice> slices = new ArrayList<>();
-        for (int sliceCount = fromSlice; sliceCount < howManySlices - fromSlice; sliceCount++) {
+        for (final int landmark : container.landmarks) {
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/cram/structure/ContainerIO.java]: Yeah I went back and forth about this ... I'll make it an int counter.

File: src/main/java/htsjdk/samtools/cram/structure/ContainerIO.java [hunk 30]
```diff
@@ -19,89 +21,97 @@
...
             slices.add(SliceIO.read(major, inputStream));
         }
 
-        container.populateSlicesAndIndexingParameters(slices);
+        container.populateSlicesAndIndexingParameters(slices, containerByteOffset);
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/cram/structure/ContainerIO.java]: Updated by removing its parameters entirely (and renaming to distributeIndexingParametersToSlices).  Slices must be now be set before calling this method.

Added tests to ContainerTest and CRAMBAiIndexerTest

File: src/main/java/htsjdk/samtools/cram/build/CramContainerHeaderIterator.java [hunk 31]
```diff
@@ -22,8 +22,19 @@ public CramContainerHeaderIterator(final InputStream inputStream) {
...
     protected Container containerFromStream(final Version cramVersion, final CountingInputStream countingStream) {
-        final Container container = ContainerIO.readContainerHeader(cramVersion.major, countingStream);
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/build/CramContainerHeaderIterator.java]: Actually, this comment is in the wrong place - it was intended for CramContainerIterator.containerFromStream, not this method.

File: src/main/java/htsjdk/samtools/CRAMBAIIndexer.java [hunk 32]
```diff
@@ -118,21 +126,21 @@ public CRAMBAIIndexer(final OutputStream output, final SAMFileHeader fileHeader)
      *
      * @param container container to be indexed
      */
-    public void processContainer(final Container container, final ValidationStringency validationStringency) {
+    void processContainer(final Container container, final ValidationStringency validationStringency) {
         if (container == null || container.isEOF()) {
             return;
         }
 
+        container.distributeIndexingParametersToSlices();
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/CRAMBAIIndexer.java]: Good catch: yes this can be removed.

There is a validation check later on, in `processAsSingleReferenceSlice()` because that is where all BAI code paths end up.

Since the indexing values all live in Slice (currently) I feel that's the appropriate place to validate them.  Maybe in the future that will change, but that's beyond the scope of this PR.

File: src/main/java/htsjdk/samtools/cram/structure/Container.java [hunk 33]
```diff
@@ -65,13 +65,55 @@
...
+
+    /**
+     * Assign {@link Slice}s to this Container.
+     * Also distribute the Container's byte offset to the {@link Slice}s, for indexing.
+     * @param slices the Slices belonging to this container
+     * @param byteOffset the byte location in the stream where this Container begins
+     */
+    void setSlices(final List<Slice> slices, final long byteOffset) {
+        this.slices = slices.toArray(new Slice[0]);
+        setByteOffsetInSlices(byteOffset);
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/cram/structure/Container.java]: sounds good

File: src/main/java/htsjdk/samtools/CRAMBAIIndexer.java [hunk 34]
```diff
@@ -296,8 +296,11 @@ public static void createIndex(final SeekableStream stream,
      * One instance is used to construct an entire index.
      * processAlignment is called for each alignment until a new reference is encountered, then
      * processReference is called when all records for the reference have been processed.
+     *
+     * TODO: this was copied from {@link htsjdk.samtools.BAMIndexer.BAMIndexBuilder}
```
Comment (cmnbroad) [inline src/main/java/htsjdk/samtools/CRAMBAIIndexer.java]: This link is broken because this class is not accessible from here; I think either just remove the `@link` or replace with `@see`.

File: src/test/java/htsjdk/samtools/cram/CRAIEntryTest.java [hunk 35]
```diff
@@ -13,13 +14,50 @@
...
+        noSize.byteSize = Slice.UNINITIALIZED_INDEXING_PARAMETER;
+        noSize.containerByteOffset = 1;
+        noSize.byteOffsetFromContainer = 1;
+
+        return new Object[][] {
+                { noContainerOffset },
+                { noOffsetFromContainer },
+                { noSize }
+        };
+    }
```
Comment (cmnbroad) [inline src/test/java/htsjdk/samtools/cram/CRAIEntryTest.java]: This code is all duplicated in `SliceTests`, and should be shared.

File: src/main/java/htsjdk/samtools/cram/structure/Slice.java [hunk 36]
```diff
@@ -113,6 +119,46 @@ public ReferenceContext getReferenceContext() {
...
+        if (index == UNINITIALIZED_INDEXING_PARAMETER) {
+            error.append("Cannot index this Slice for BAI because its index is unknown.").append(System.lineSeparator());
+        }
+
+        if (error.length() > 0) {
+            throw new CRAMException(error.toString());
+        }
+    }
+
+    public void craiIndexInitializationCheck() {
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/cram/structure/Slice.java]: I'll add a comment to make the difference clearer.

I'd prefer to keep them separate, since they both test for exactly what is needed under their own requirements.

File: src/main/java/htsjdk/samtools/CRAMBAIIndexer.java [hunk 37]
```diff
@@ -61,11 +62,18 @@
...
  * Class for both constructing BAM index content and writing it out.
+ *
  * There are two usage patterns:
- * 1) Building a bam index from an existing cram file
- * 2) Building a bam index while building the cram file
- * In both cases, processAlignment is called for each cram slice and
- * finish() is called at the end.
+ * 1) Building a bam index (BAI) while building the CRAM file
+ * 2) Building a bam index (BAI) from an existing CRAI file
+ *
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply src/main/java/htsjdk/samtools/CRAMBAIIndexer.java]: Right - that's why I didn't include it.  I can add it here with a note that it's unused.

File: src/test/java/htsjdk/samtools/cram/CRAIEntryTest.java (hunk 35 above)
Comment (jmthibault79) [reply src/test/java/htsjdk/samtools/cram/CRAIEntryTest.java]: Good call.

File: src/main/java/htsjdk/samtools/cram/structure/Slice.java (hunk 36 above)
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/cram/structure/Slice.java]: I don't see why the distinction, or the increased complexity and code duplication, is useful - there is only one `distributeIndexingParametersToSlices` thats establishes these values. But ok.
